### PR TITLE
Identity 3200 - Use SQL LCASE() to compare AUTHZ_USER in IDN_OAUTH2_ACCESS_TOKEN

### PR DESCRIPTION
**Pull request by madurangasiriwardena**

### CODE REVIEW

File: components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/AbstractAuthorizationGrantHandler.java [hunk 1]
```diff
@@ -94,7 +94,14 @@ public OAuth2AccessTokenRespDTO issue(OAuthTokenReqMessageContext tokReqMsgCtx)
 
         String consumerKey = tokReqMsgCtx.getOauth2AccessTokenReqDTO().getClientId();
         String authorizedUser = tokReqMsgCtx.getAuthorizedUser();
-        OAuthCacheKey cacheKey = new OAuthCacheKey(consumerKey + ":" + authorizedUser.toLowerCase() + ":" + scope);
+        boolean isUsernameCaseSensitive = OAuth2Util.isUsernameCaseSensitive(authorizedUser);
+        String cacheKeyString;
+        if (isUsernameCaseSensitive){
+            cacheKeyString = consumerKey + ":" + authorizedUser + ":" + scope;
+        }else {
```
Comment (pulasthi7) [inline components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/AbstractAuthorizationGrantHandler.java]: format the code

File: components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/RefreshGrantHandler.java [hunk 2]
```diff
@@ -248,7 +248,15 @@ public OAuth2AccessTokenRespDTO issue(OAuthTokenReqMessageContext tokReqMsgCtx)
         // if it's enabled.
         if (cacheEnabled) {
             // Remove the old access token from the OAuthCache
-            CacheKey oauthCacheKey = new OAuthCacheKey(clientId + ":" + authorizedUser + ":" + scope);
+            boolean isUsernameCaseSensitive = OAuth2Util.isUsernameCaseSensitive(authorizedUser);
+            String cacheKeyString;
+            if (isUsernameCaseSensitive){
+                cacheKeyString = clientId + ":" + authorizedUser + ":" + scope;
+            }else {
```
Comment (pulasthi7) [inline components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/RefreshGrantHandler.java]: format the code

File: components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/OAuth2Util.java [hunk 3]
```diff
@@ -252,8 +261,15 @@ public static String getAuthenticatedUsername(String clientId, String clientSecr
                  * own cache key and cache entry class every time we need to put something to it? Ideal solution is
                  * to have a generalized way of caching a key:value pair
                  */
-                cache.addToCache(new OAuthCacheKey(clientId + ":" + username), new ClientCredentialDO(username));
-                log.debug("Caching username : " + username);
+                if (isUsernameCaseSensitive){
+                    cache.addToCache(new OAuthCacheKey(clientId + ":" + username), new ClientCredentialDO(username));
+                }else {
```
Comment (pulasthi7) [inline components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/OAuth2Util.java]: format the code

File: components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/DefaultIDTokenBuilder.java [hunk 4]
```diff
@@ -287,8 +287,14 @@ private long getAccessTokenIssuedTime(String accessToken, OAuthTokenReqMessageCo
         TokenMgtDAO tokenMgtDAO = new TokenMgtDAO();
 
         OAuthCache oauthCache = OAuthCache.getInstance();
+        String authorizedUser = request.getAuthorizedUser();
+        boolean isUsernameCaseSensitive = OAuth2Util.isUsernameCaseSensitive(authorizedUser);
+        if (!isUsernameCaseSensitive){
```
Comment (omindu) [inline components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/DefaultIDTokenBuilder.java]: Formatting issue.

File: components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/TokenMgtDAO.java [hunk 5]
```diff
@@ -910,8 +932,15 @@ public void revokeToken(String tokenId, String userId) throws IdentityOAuth2Exce
             }
             String sqlQuery = SQLQueries.GET_DISTINCT_APPS_AUTHORIZED_BY_USER_ALL_TIME.replace(
                     IDN_OAUTH2_ACCESS_TOKEN, accessTokenStoreTable);
+            if (!isUsernameCaseSensitive){
```
Comment (omindu) [inline components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/TokenMgtDAO.java]: Formatting issue.

File: components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/TokenMgtDAO.java [hunk 6]
```diff
@@ -390,9 +403,17 @@ public AccessTokenDO retrieveLatestAccessToken(String consumerKey, String userNa
             if (StringUtils.isNotEmpty(userStoreDomain)) {
                 sql = sql.replace(IDN_OAUTH2_ACCESS_TOKEN, IDN_OAUTH2_ACCESS_TOKEN + "_" + userStoreDomain);
             }
+            if (!isUsernameCaseSensitive){
```
Comment (omindu) [inline components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/TokenMgtDAO.java]: Formatting issue

File: components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/dao/OAuthAppDAO.java [hunk 7]
```diff
@@ -348,11 +361,20 @@ private boolean isDuplicateApplication(String username, int tenantId, OAuthAppDO
...
-            prepStmt = connection.prepareStatement(SQLQueries.OAuthAppDAOSQLQueries.CHECK_EXISTING_APPLICATION);
-            prepStmt.setString(1, username.toLowerCase());
+            String sql = SQLQueries.OAuthAppDAOSQLQueries.CHECK_EXISTING_APPLICATION;
+            if (!isUsernameCaseSensitive){
+                sql.replace("USERNAME", "LOWER(USERNAME)");
+            }
+            prepStmt = connection.prepareStatement(sql);
+            if (isUsernameCaseSensitive){
+                prepStmt.setString(1, username);
+            }else {
```
Comment (omindu) [inline components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/dao/OAuthAppDAO.java]: Formatting issue.

File: components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/dao/OAuthAppDAO.java [hunk 8]
```diff
@@ -138,16 +139,28 @@ public void addOAuthApplication(OAuthAppDO consumerAppDO) throws IdentityOAuthAd
...
             connection = JDBCPersistenceManager.getInstance().getDBConnection();
-            prepStmt = connection.prepareStatement(SQLQueries.OAuthAppDAOSQLQueries.GET_APPS_OF_USER_WITH_TENANTAWARE_OR_TENANTUNAWARE_USERNAME);
-            prepStmt.setString(1, tenantAwareUserName);
-            prepStmt.setString(2, tenantUnawareUserName);
+            String sql = SQLQueries.OAuthAppDAOSQLQueries.GET_APPS_OF_USER_WITH_TENANTAWARE_OR_TENANTUNAWARE_USERNAME;
+            if (!isUsernameCaseSensitive){
+                sql.replace("USERNAME", "LOWER(USERNAME)");
+            }
+            prepStmt = connection.prepareStatement(sql);
+            if (isUsernameCaseSensitive){
```
Comment (omindu) [inline components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/dao/OAuthAppDAO.java]: Formatting issue.

**Comment by omindu** [comment]
Please check the PR for for formatting issues. There are a number of places with missing spaces.

File: components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/AbstractAuthorizationGrantHandler.java (hunk 1 above)
Comment (madurangasiriwardena) [inline components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/AbstractAuthorizationGrantHandler.java]: Fixed with https://github.com/wso2/carbon-identity/pull/1679

File: components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/RefreshGrantHandler.java (hunk 2 above)
Comment (madurangasiriwardena) [inline components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/RefreshGrantHandler.java]: Fixed with https://github.com/wso2/carbon-identity/pull/1679

File: components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/OAuth2Util.java (hunk 3 above)
Comment (madurangasiriwardena) [inline components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/OAuth2Util.java]: Fixed with https://github.com/wso2/carbon-identity/pull/1679

File: components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/DefaultIDTokenBuilder.java (hunk 4 above)
Comment (madurangasiriwardena) [inline components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/DefaultIDTokenBuilder.java]: Fixed with https://github.com/wso2/carbon-identity/pull/1679

File: components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/TokenMgtDAO.java (hunk 5 above)
Comment (madurangasiriwardena) [inline components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/TokenMgtDAO.java]: Fixed with https://github.com/wso2/carbon-identity/pull/1679

File: components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/TokenMgtDAO.java (hunk 6 above)
Comment (madurangasiriwardena) [inline components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/TokenMgtDAO.java]: Fixed with https://github.com/wso2/carbon-identity/pull/1679

File: components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/dao/OAuthAppDAO.java (hunk 7 above)
Comment (madurangasiriwardena) [inline components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/dao/OAuthAppDAO.java]: Fixed with https://github.com/wso2/carbon-identity/pull/1679

File: components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/dao/OAuthAppDAO.java (hunk 8 above)
Comment (madurangasiriwardena) [inline components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/dao/OAuthAppDAO.java]: Fixed with https://github.com/wso2/carbon-identity/pull/1679

**Pull request merged**